oppm_linker/oppm_linker.man
## NAME
  oppm_linker - rc daemon for automatically linking files in a Git repository to their install location

## SYNOPSIS
  `rc oppm_linker enable`
  
  `rc oppm_linker disable`
  
  `rc oppm_linker start [OPTION]...`

## DESCRIPTION
  The oppm_linker aims to solve the problem with keeping in-development programs on an OpenComputers machine synchronized with files in a Git repository. This is done by having the user clone a Git repository into the OC machine, adding a `programs.cfg` file to define packages and their install locations, then oppm_linker creates symbolic links from the original files in the places where they would be installed. This simulates what would happen if the packages were all installed with OPPM, but keeps the original files in the Git repo without duplicating them. Since symlinks in OC do not persist across a reboot, oppm_linker runs once at startup to create the symlinks.

### Setup instructions:
  1. First consider setting `bufferChanges=false` and increasing the value of `hddSizes` for the hard disk you want to use in the OpenComputers configuration. The `bufferChanges` setting prevents the in-game file systems from writing their contents to disk until the world is saved, and this is a problem if using an external editor to write programs. The default hard disk sizes are also quite small and may not fit the whole Git repo (however a shallow clone of the repo may work).
  
  2. Get OpenOS and OPPM installed on the OC machine (you can craft the floppy disk to install OPPM).
  
  3. Run `mkdir /repository` (or `mkdir /repo`, either one works) in-game, then copy the Git repo into here outside of the game or `git clone` one from GitHub (use `--depth 1` for a shallow clone to save disk space). The root of the repo should be at `/repository/<your-repo-name>/.git`. You can add as many repos in `/repository` as you want and they will all get loaded by oppm_linker.
  
      - Another option is the OPPM package called `gitrepo` to clone a repo from within the game, or just download a zip archive from the internet with `wget`.
      
      - If using a third-party launcher for the game (such as Prism Launcher) on Windows, you may run into MAX_PATH issues when cloning. In the Prism settings I moved my instances folder into my home directory and this reduced the filename lengths enough for the clone to succeed.
  
  4. The Git repo should have a valid `programs.cfg` file to define the packages. See https://ocdoc.cil.li/tutorial:program:oppm for details.
  
  5. Get oppm_linker installed with `oppm install oppm_linker`, then enable the daemon to run at boot with `rc oppm_linker enable`. It's important for oppm_linker to be the first daemon to run on the system since it sets up binaries, so it should appear as the first entry in `/etc/rc.cfg`. After a reboot (it's usually best to cold-reboot the system by pressing power button), the new symlinks should show up in the package install locations. Note that the symlinks will not show in the physical filesystem outside the game.
  
  6. If the verbose output from creating symlinks is annoying, add the line `oppm_linker = "-s"` to `/etc/rc.cfg`. This enables the silent option and only errors will show.
  
  To refresh the symlinks if a new package was added or files changed, you can manually run `rc oppm_linker start`. This also helps to debug errors and warnings that may be showing during boot. Note that oppm_linker will not delete any old symlinks (they are removed after a reboot anyways) and will not delete directories that no longer exist in the repo.
  
  One more useful tip: creating copies of the hard disk that was used to set this up (using cheats) can be helpful to keep the same files synchronized across multiple computers. This can also be done by setting up a RAID and cloning the block to anywhere it is needed. Multiple computers can also access a single RAID when cheats are not an option.

## OPTIONS
  `-s`  suppress all normal output (except for errors)
  
  `-f`  overwrite existing files and create missing directories

## EXAMPLES
  `rc oppm_linker start -sf`
  > Update symlinks silently and force existing files to be replaced.

## SEE ALSO
  [oppm](https://github.com/OpenPrograms/Vexatos-Programs/tree/master/oppm)


oppm_linker/oppm_linker.lua
--------------------------------------------------------------------------------
-- An rc daemon for automatically linking files in a Git repository to their
-- install location. This is designed for easy development of code with the
-- OpenPrograms Package Manager.
-- 
-- @see file://oppm_linker/README.md
-- @author tdepke2
--------------------------------------------------------------------------------


local component = require("component")
local computer = require("computer")
local fs = require("filesystem")
local serialization = require("serialization")
local shell = require("shell")
local text = require("text")

-- Time in seconds to pause if a warning shows up and not running in silent mode.
local WARNING_PAUSE_TIME = 4

local numWarnings = 0


-- Read the configuration file used by OPPM. This is based on the same process
-- OPPM uses to load in this file. Note that this is currently only used to grab
-- the default install path, any additional repos defined in the config are
-- ignored.
local function readOppmConfig()
  local filename = "/etc/oppm.cfg"
  if not fs.exists(filename) then
    filename = fs.concat(fs.path(shell.resolve(os.getenv("_"))), "/etc/oppm.cfg")
  end
  if not fs.exists(filename) then
    return {-1}
  end
  local file, errMessage = io.open(filename, "rb")
  assert(file, "failed to open file \"" .. filename .. "\": " .. tostring(errMessage))
  local cfgTable = serialization.unserialize(file:read("*a"))
  file:close()
  return cfgTable or {-1}
end


-- Creates a symbolic link from srcPath to destPath. Additionally does some
-- error checking to make sure we don't create any broken links and the
-- directories all exist.
local function addLink(options, srcPath, destPath)
  assert(fs.exists(srcPath), "file at \"" .. srcPath .. "\" not found")
  local destFilename = fs.concat(destPath, fs.name(srcPath))
  if fs.exists(destFilename) then
    if fs.isLink(destFilename) or options.force then
      assert(fs.remove(destFilename))
    else
      assert(false, "file at \"" .. destFilename .. "\" already exists (use \'-f\' flag to remove it)")
    end
  elseif not fs.exists(destPath) then
    assert(options.force, "path to \"" .. destPath .. "\" does not exist (use \'-f\' flag to create it)")
    if not options.silent then
      io.write("Creating directory \"", destPath, "\".\n")
    end
    assert(fs.makeDirectory(destPath))
  end
  if not options.silent then
    io.write("Linking \"", srcPath, "\" to \"", destFilename, "\".\n")
  end
  assert(fs.link(srcPath, destFilename))
end


-- Read the contents of the programs.cfg file for the repository at repoPath,
-- and add symlinks to all the specified files. If a repository doesn't have a
-- programs.cfg file we send a warning and move on.
local function readPackageConfig(options, repoPath, installPath)
  local filename = fs.concat(repoPath, "/programs.cfg")
  if not fs.exists(filename) then
    if not options.silent then
      io.write("\27[33mWarning: missing package definitions at \"", filename, "\"\27[0m\n")
      numWarnings = numWarnings + 1
    end
    return
  end
  local file, errMessage = io.open(filename, "rb")
  assert(file, "failed to open file \"" .. filename .. "\": " .. tostring(errMessage))

  local pkgTable = serialization.unserialize(file:read("*a"))
  file:close()
  assert(pkgTable, "failed to deserialize programs list \"" .. filename .. "\"")

  -- The programs.cfg file just gives a table with package names, where each one defines files in the repo and their install path.
  for pkgName, pkgDat in pairs(pkgTable) do
    -- Special case since the oppm_linker files need to be persistent on the system and we don't want to try to create symlinks to replace them.
    if pkgName == "oppm_linker" then
      pkgDat.files = nil
    end
    for srcPath, destPath in pairs(pkgDat.files or {}) do
      -- A colon at start of srcPath means all of the contents in that directory. A question mark has another meaning at start of srcPath but we ignore it.
      local addContents = string.find(srcPath, "^:")
      srcPath = fs.concat(repoPath, string.match(srcPath, "[^:?](/.*)"))

      -- Double slash at start of destPath indicates an absolute path, otherwise we use the installPath.
      if string.find(destPath, "^//") then
        destPath = string.sub(destPath, 2)
      else
        destPath = fs.concat(installPath, destPath)
      end

      if addContents then
        local pathIter, errMessage = fs.list(srcPath)
        assert(pathIter, "failed to list directory \"" .. srcPath .. "\": " .. tostring(errMessage))
        for filename in pathIter do
          addLink(options, fs.concat(srcPath, filename), destPath)
        end
      else
        addLink(options, srcPath, destPath)
      end
    end
  end
end


-- Search for all of the repositories in "/repository" or "/repo" and process
-- the OPPM packages file in each one.
local function main(options)
  local installPath = readOppmConfig().path or "/usr"

  local repoRoot
  if fs.exists("/repository") then
    repoRoot = "/repository"
  elseif fs.exists("/repo") then
    repoRoot = "/repo"
  else
    assert(false, "no directory \"/repository\" or \"/repo\" found.")
  end

  local pathIter, errMessage = fs.list(repoRoot)
  assert(pathIter, "failed to list directory \"" .. repoRoot .. "\": " .. tostring(errMessage))
  for filename in pathIter do
    readPackageConfig(options, fs.concat(repoRoot, filename), installPath)
  end
end


-- Called by rc when daemon runs the first time, or when manually invoked.
function start(...)
  -- Parse arguments from the rc config if during startup, or from the function arguments if user invoked the start function.
  local arguments, options
  local isStarting = (computer.runlevel() == "S")
  if isStarting then
    arguments, options = shell.parse(table.unpack(text.tokenize(args or "")))
  else
    arguments, options = shell.parse(...)
  end
  options = {
    force = options.f,
    silent = options.s
  }

  if not options.silent then
    require("term").clear()
  end

  -- Run main function in a protected call. Pause system and print errors if something went wrong.
  local status, err = pcall(main, options)
  if not status then
    if options.silent then
      require("term").clear()
    end
    io.write("\27[31mERROR: failed to run oppm_linker\n")
    io.write(tostring(err), "\n")
    if isStarting then
      io.write("\27[0m\n(press enter to continue)\n")
      io.read()
    end
  elseif numWarnings > 0 then
    io.write("\27[33moppm_linker: ", numWarnings, " warning(s) generated.\27[0m\n")
    if isStarting and WARNING_PAUSE_TIME > 0 then
      os.sleep(WARNING_PAUSE_TIME)
    end
    numWarnings = 0
  end
end
